module/zfs/dnode.c

```diff
@@ -1062,36 +1066,26 @@ dnode_buf_evict_async(void *dbu)
  *
  * The dnode_phys_t buffer may not be in sync with the in-core dnode
  * structure, so we try to check the dnode structure first and fall back
- * to the dnode_phys_t buffer it doesn't exist.
+ * to the dnode_phys_t buffer it doesn't exist.  When an in-code dnode
+ * exists we can always trust dn->dn_num_slots to be accurate, even for
+ * a held dnode which has not yet been fully allocated.
  */
 static boolean_t
-dnode_is_consumed(dmu_buf_impl_t *db, int idx)
+dnode_is_consumed(dnode_children_t *children, dnode_phys_t *dn_block, int idx)
 {
-	dnode_handle_t *dnh;
-	dmu_object_type_t ot;
-	dnode_children_t *children_dnodes;
-	dnode_phys_t *dn_block;
-	int skip;
-	int i;
-
-	children_dnodes = dmu_buf_get_user(&db->db);
-	dn_block = (dnode_phys_t *)db->db.db_data;
+	int skip, i;
 
 	for (i = 0; i < idx; i += skip) {
-		dnh = &children_dnodes->dnc_children[i];
+		dnode_handle_t *dnh = &children->dnc_children[i];
 
-		zrl_add(&dnh->dnh_zrlock);
 		if (dnh->dnh_dnode != NULL) {
-			ot = dnh->dnh_dnode->dn_type;
 			skip = dnh->dnh_dnode->dn_num_slots;
 		} else {
-			ot = dn_block[i].dn_type;
-			skip = dn_block[i].dn_extra_slots + 1;
+			if (dn_block[i].dn_type != DMU_OT_NONE)
+				skip = dn_block[i].dn_extra_slots + 1;
```

Comment:
Is it necessary for dnode_destroy() to zero dn_type and dn_extra_slots so that holes created by object destruction do not leave nonzero values that result in skip getting an incorrect value when children[] is out of sync with the dbuf and dnh_dnode == NULL?